src/librustc/infer/freshen.rs

```diff
@@ -32,15 +32,16 @@
 
 use ty::{self, Ty, TyCtxt, TypeFoldable};
 use ty::fold::TypeFolder;
-use std::collections::hash_map::{self, Entry};
+use util::nodemap::FnvHashMap;
+use std::collections::hash_map::Entry;
 
 use super::InferCtxt;
 use super::unify_key::ToType;
 
 pub struct TypeFreshener<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
     infcx: &'a InferCtxt<'a, 'gcx, 'tcx>,
     freshen_count: u32,
-    freshen_map: hash_map::HashMap<ty::InferTy, Ty<'tcx>>,
```

Comment:
Some tests on `syntex_syntax`:

With this PR:
```
time: 0.557; rss: 230MB item-types checking
time: 5.720; rss: 262MB item-bodies checking

time: 0.574; rss: 230MB item-types checking
time: 5.779; rss: 262MB item-bodies checking

time: 0.559; rss: 231MB item-types checking
time: 5.761; rss: 261MB item-bodies checking

time: 0.566; rss: 232MB item-types checking
time: 5.708; rss: 263MB item-bodies checking
```

Without this PR (current master):
```
time: 0.560; rss: 232MB item-types checking
time: 5.768; rss: 263MB item-bodies checking

time: 0.565; rss: 230MB item-types checking
time: 5.907; rss: 261MB item-bodies checking

time: 0.561; rss: 230MB item-types checking
time: 5.810; rss: 261MB item-bodies checking

time: 0.557; rss: 231MB item-types checking
time: 5.812; rss: 261MB item-bodies checking
```